pkg/monitor/controller_test.go

```diff
@@ -0,0 +1,682 @@
+
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/apis/crd/v1beta1"
+	fakeclientset "antrea.io/antrea/pkg/client/clientset/versioned/fake"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions"
+	"antrea.io/antrea/pkg/controller/grouping"
+	"antrea.io/antrea/pkg/controller/labelidentity"
+	"antrea.io/antrea/pkg/controller/networkpolicy"
+	"antrea.io/antrea/pkg/controller/networkpolicy/store"
+	"antrea.io/antrea/pkg/controller/querier"
+)
+
+const (
+	informerDefaultResync = 12 * time.Second
+)
+
+type fakeController struct {
+	controllerMonitor  *controllerMonitor
+	crdClient          *fakeclientset.Clientset
+	client             *fake.Clientset
+	informerFactory    informers.SharedInformerFactory
+	crdInformerFactory crdinformers.SharedInformerFactory
+}
+
+func newControllerMonitor(crdClient *fakeclientset.Clientset) *fakeController {
+	client := fake.NewSimpleClientset()
+	informerFactory := informers.NewSharedInformerFactory(client, informerDefaultResync)
+	crdInformerFactory := crdinformers.NewSharedInformerFactory(crdClient, informerDefaultResync)
+	namespaceInformer := informerFactory.Core().V1().Namespaces()
+	serviceInformer := informerFactory.Core().V1().Services()
+	networkPolicyInformer := informerFactory.Networking().V1().NetworkPolicies()
+	nodeInformer := informerFactory.Core().V1().Nodes()
+	cnpInformer := crdInformerFactory.Crd().V1alpha1().ClusterNetworkPolicies()
+	anpInformer := crdInformerFactory.Crd().V1alpha1().NetworkPolicies()
+	tierInformer := crdInformerFactory.Crd().V1alpha1().Tiers()
+	cgInformer := crdInformerFactory.Crd().V1alpha3().ClusterGroups()
+	grpInformer := crdInformerFactory.Crd().V1alpha3().Groups()
+	externalNodeInformer := crdInformerFactory.Crd().V1alpha1().ExternalNodes()
+
+	addressGroupStore := store.NewAddressGroupStore()
+	appliedToGroupStore := store.NewAppliedToGroupStore()
+	networkPolicyStore := store.NewNetworkPolicyStore()
+	groupStore := store.NewGroupStore()
+	groupEntityIndex := grouping.NewGroupEntityIndex()
+	labelIdentityIndex := labelidentity.NewLabelIdentityIndex()
+
+	networkPolicyController := networkpolicy.NewNetworkPolicyController(client,
+		crdClient,
+		groupEntityIndex,
+		labelIdentityIndex,
+		namespaceInformer,
+		serviceInformer,
+		networkPolicyInformer,
+		nodeInformer,
+		cnpInformer,
+		anpInformer,
+		tierInformer,
+		cgInformer,
+		grpInformer,
+		addressGroupStore,
+		appliedToGroupStore,
+		networkPolicyStore,
+		groupStore,
+		false)
+
+	controllerQuerier := querier.NewControllerQuerier(networkPolicyController, 10349)
+	externalNodeEnabled := true
+	controllerMonitor := NewControllerMonitor(crdClient, nodeInformer, externalNodeInformer, controllerQuerier, externalNodeEnabled)
+
+	return &fakeController{
+		controllerMonitor,
+		crdClient,
+		client,
+		informerFactory,
+		crdInformerFactory,
+	}
+}
+
+func initController(controller *fakeController) {
+	stopCh := make(chan struct{})
+	defer close(stopCh)
```

Comment:
I don't understand this. Do you mean to close `stopCh` when this function returns? This doesn't seem correct and I would be surprised if the tests were not flaky. I don't think the informer factories should be stopped before the end of each test. You can pass a `stopCh` as a parameter to the function instead.

Reply:
fixed